## Second Assignment


Consider the data from the Spanish electrical system used in class or any other database of your interest including one dependent variable $Y$ and at least 5 independent variables $X$, with a minimum of 500 observations. We seek to adjust a multiple linear regression model to explain variable $Y$ as a function of the other variables $X$, i.e., ($Y = X \beta + \varepsilon $) by using "Ridge Regression":

$$
\min \quad ||Y-X\beta||_2^2 + \rho || \beta ||_2^2 
$$

where $\rho$ is a parameter.

I considered the data from the Spanish electrical system in order to solve the problem and a value of $\rho=1$.

**a)** Estimate the value of the regression coefficients by implementing the analytical
solution.

**SOLUTION**

Let's consider the derivative of the function respect to $\beta$ (note that the square of the norm ease the calculation of the derivative, because we see it as product of vectors):

$\frac{\delta}{\delta \beta}((Y - X\beta)'(Y-X\beta) + \rho \beta'\beta))=-2X'(Y-\beta' X) + 2\rho\beta = 0 \Rightarrow X'Y = X'X\beta + \rho \beta \Rightarrow \beta = (X'X+ \rho I )^{-1}X' Y$

Note that the function is convex (sum of convex is convex), so $\beta = (X'X+ \rho I )^{-1}X' Y$ is the global minimum of the funcion.

Now applying the obtained formula for $\beta$ to our particular problem:

In [75]:
%matplotlib inline
import numpy as np
import time
import pandas as pd
import matplotlib.pyplot as plt

price=pd.read_csv('price.csv',delimiter=';',usecols=['value'])
coal=pd.read_csv('coal.csv',delimiter=';',usecols=['value'])
gas=pd.read_csv('gas.csv',delimiter=';',usecols=['value'])
wind=pd.read_csv('wind.csv',delimiter=';',usecols=['value'])
hidro=pd.read_csv('hidro.csv',delimiter=';',usecols=['value'])
nuclear=pd.read_csv('nuclear.csv',delimiter=';',usecols=['value'])
solar=pd.read_csv('solar.csv',delimiter=';',usecols=['value'])
cogen=pd.read_csv('cogen.csv',delimiter=';',usecols=['value'])
demand=pd.read_csv('demand.csv',delimiter=';',usecols=['value'])
df=pd.concat([price, coal, gas, wind, hidro, nuclear, solar, cogen, demand],axis=1)
df.columns=['price', 'coal', 'gas', 'wind', 'hidro', 'nuclear', 'solar', 'cogen', 'demand']

normalized_df=(df-df.mean())/df.std()
normalized_df

price      coal       gas      wind     hidro   nuclear     solar  \
0    -0.843676 -0.863129 -1.032348  0.429243 -0.291410  1.504478 -0.462052   
1    -1.140834 -1.183432 -1.145887  0.464282 -0.526312  1.510829 -0.488742   
2    -1.486618 -1.435346 -1.258767  0.469869 -0.952336  1.515959 -0.546745   
3    -1.609803 -1.498452 -1.278001  0.372891 -0.936070  1.516936 -0.588177   
4    -1.655187 -1.509630 -1.271681  0.293776 -0.890256  1.523287 -0.684336   
5    -1.609803 -1.420612 -1.206778  0.146101 -0.859893  1.524752 -0.754094   
6    -1.140834 -1.262187 -1.099614  0.024636 -0.297103  1.520355 -0.769414   
7    -0.580016 -1.034357 -0.908422 -0.094622  0.106962  1.523775 -0.765367   
8    -0.359579 -1.029276 -0.741961 -0.189876  0.500455  1.524997 -0.630379   
9    -0.303389 -0.854288 -0.556649 -0.397835  1.296386  1.525729 -0.103624   
10   -0.386593 -1.059762 -0.579126 -0.458464  1.496859  1.524264  0.912790   
11   -0.360660 -1.124392 -0.603142 -0.354933  1.124106  1.516447  1.636390   
12   -0.277455 -1.086284 -0.601493 -0.318790  1.216549  1.520355  1.937489   
13   -0.225588 -1.022569 -0.572147 -0.272922  1.181171  1.513272  2.042127   
14   -0.386593 -1.074192 -0.636665 -0.083310  0.594660  1.513760  2.096181   
15   -0.603789 -1.328849 -0.756744  0.262324  0.100321  1.512783  2.006381   
16   -0.827468 -1.333219 -0.859456  0.608302 -0.014759  1.507165  1.831888   
17   -1.016568 -1.381996 -0.901883  0.976076 -0.148543  1.508142  1.573473   
18   -1.140834 -1.502009 -1.036415  1.378614 -0.308353  1.510096  1.200881   
19   -1.133270 -1.526804 -0.992615  1.684932 -0.379515  1.511806  0.601381   
20   -0.654576 -1.474775 -0.917050  1.775082  0.064943  1.517424 -0.131759   
21   -0.602708 -1.318078 -0.817085  1.865646  0.300117  1.518157 -0.408770   
22   -1.286712 -1.530360 -1.084941  1.818674  0.113604  1.515226 -0.444131   
23   -1.769728 -1.700166 -1.245192  1.621130 -0.070197  1.503012 -0.451068   
24   -1.790259 -1.887247 -1.240411  1.533670 -0.457997  1.505699 -0.480455   
25   -1.891833 -2.038456 -1.191171  1.513598 -0.865044  1.512539 -0.530655   
26   -1.978279 -2.015897 -1.201557  1.079608 -1.362636  1.515470 -0.581336   
27   -2.102545 -2.027176 -1.147865  0.897721 -1.608789  1.516203 -0.642808   
28   -2.136043 -2.069450 -1.263987  0.953246 -1.686051  1.516691 -0.670846   
29   -2.123076 -2.070466 -1.192545  0.871234 -1.663144  1.521088 -0.753998   
...        ...       ...       ...       ...       ...       ...       ...   
2155  2.287828  1.609469  2.112013  0.781084  2.120239 -1.129491 -0.780784   
2156  2.322406  1.663429  2.231597  0.881857  1.789505 -1.128025 -0.814218   
2157  2.345098  1.704991  1.941979  0.938416  1.700180 -1.127536 -0.815181   
2158  1.509814  1.733445  1.384945  1.073676  0.137460 -1.122407 -0.815470   
2159  0.978172  1.706414  0.922931  1.006425 -0.371111 -1.121185 -0.815470   
2160  0.794474  1.639244  0.297037  0.947107 -0.441053 -1.122162 -0.815470   
2161  0.365486  1.645849 -0.507463  0.771980 -0.650744 -1.121185 -0.815470   
2162  0.068328  1.702857 -0.692445  0.700384 -0.684224 -1.124361 -0.815470   
2163 -0.125094  1.601645 -0.721297  0.702246 -1.563381 -1.122162 -0.815470   
2164 -0.299067  1.522382 -0.731574  0.660999 -2.190963 -1.114590 -0.815663   
2165 -0.319598  1.539454 -0.594349  0.560020 -2.292487 -1.114590 -0.815663   
2166 -0.092677  1.648898 -0.476908  0.522428 -1.147388 -1.119720 -0.815470   
2167  0.954399  1.681212 -0.120573  0.558295 -0.417875 -1.122407 -0.815470   
2168  2.131144  1.737814  1.260910  0.583057  0.042306 -1.120697 -0.815567   
2169  2.305117  1.776531  1.979734  0.497735  0.585307 -1.126315 -0.764789   
2170  2.395885  1.785677  1.992099  0.376063  1.251926 -1.134864 -0.253066   
2171  2.395885  1.790961  2.061563  0.252736  1.422173 -1.139994  0.430935   
2172  2.337534  1.786388  1.998804  0.273980  0.994523 -1.137307  0.912597   
2173  2.305117  1.783949  2.031887  0.393376  0.282902 -1.128758  1.112527   
2174  2.305117  1.783238  1.988

In [76]:
[a,b]=df.shape
ntest=100
df_train = normalized_df.iloc[:(a-ntest), :]
df_test = normalized_df.iloc[(a-ntest):, :]
# train
Y = np.array(df_train['price'])
Y = np.expand_dims(Y, axis=1)
X0 = np.ones([a-ntest,1])
X1 = np.array(df_train[df_train.columns[df_train.columns!='price']])
X = np.concatenate([X0, X1],axis=1)
#test
Y_test = np.array(df_test['price'])
Y_test = np.expand_dims(Y_test, axis=1)
X0_test = np.ones([ntest,1])
X1_test= np.array(df_test[df_test.columns[df_test.columns!='price']])
X_test = np.concatenate([X0_test, X1_test],axis=1)

In [77]:
rho=1
from numpy.linalg import inv
time_start = time.clock()
beta_ls_exact= np.dot( np.dot(inv(np.dot(X.T,X) +rho*np.identity(b) ),X.T),Y)
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print(beta_ls_exact)

('time elapsed=', 0.0004329874827817548)
[[-0.0115803 ]
 [ 0.36259694]
 [ 0.21770513]
 [-0.06965771]
 [ 0.36784859]
 [-0.14696702]
 [-0.00694018]
 [ 0.07745367]
 [-0.04210075]]


**b)** Estimate the value of the regression coeffcients by using the function minimize
from the Python module Scipy.optimize. Try at least four available solvers and compare
their performance in terms of number of iterations, number of function, gradient and hessian
evaluations as well as total computational time.

In [7]:
help(minimize)

Help on function minimize in module scipy.optimize._minimize:

minimize(fun, x0, args=(), method=None, jac=None, hess=None, hessp=None, bounds=None, constraints=(), tol=None, callback=None, options=None)
    Minimization of scalar function of one or more variables.
    
    In general, the optimization problems are of the form::
    
        minimize f(x) subject to
    
        g_i(x) >= 0,  i = 1,...,m
        h_j(x)  = 0,  j = 1,...,p
    
    where x is a vector of one or more variables.
    ``g_i(x)`` are the inequality constraints.
    ``h_j(x)`` are the equality constrains.
    
    Optionally, the lower and upper bounds for each element in x can also be
    specified using the `bounds` argument.
    
    Parameters
    ----------
    fun : callable
        Objective function.
    x0 : ndarray
        Initial guess.
    args : tuple, optional
        Extra arguments passed to the objective function and its
        derivatives (Jacobian, Hessian).
    method : str or callable, op

In [4]:
from scipy.optimize import minimize

We are going to use the solvers and see which is the best for us:
- 'Nelder-Mead'
- 'Powell'
- 'CG'
- 'BFGS'
- 'Newton-CG'
- 'L-BFGS-B'   
- 'TNC'    
- 'COBYLA'  
- 'SLSQP' 
- 'dogleg'
- 'trust-ncg'

In [5]:
#Here we have the funtion that we want to minimize
def ridge_reg(beta_ls, X, Y,rho):
    beta_ls=np.matrix(beta_ls)
    z=Y-np.dot(X,beta_ls.T)
    return np.dot(z.T,z) + rho*np.dot(beta_ls,beta_ls.T)

We are going to need also the expression of the gradien and the hessian.

**Gradient**: $ -2X' Y+2(X' X+\rho I)\beta $


**Hessian**: $2(X'X+\rho\,I)$

In [6]:
#Gradient:
def ridge_reg_der(beta_ls,X,Y,rho):
    beta_ls=np.matrix(beta_ls)
    pp=-2*np.dot((Y-np.dot(X,beta_ls.T)).T,X)+2*rho*beta_ls
    aa=np.squeeze(np.asarray(pp))
    return aa

#Hessian:
def ridge_reg_hess(beta_ls,X,Y,rho):
    return 2*np.dot(X.T,X)+2*rho*np.identity(b)


## 'Nelder-Mead'

In [20]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Nelder-Mead",
                            options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))


('time elapsed=', 0.15454097598382077)
[-0.04500509  0.31478912  0.12121656 -0.14988045  0.31996648 -0.24497396
 -0.07880544 -0.00367904  0.13425021]
('error=', 0.46041063550883704)


The method did not converge.

In [21]:
res.keys()

['status', 'success', 'final_simplex', 'nfev', 'fun', 'x', 'message', 'nit']

In [23]:
print res.nit 
print res.fun

1236
406.087802983


## 'Powell'

In [24]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Powell",options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.679092
         Iterations: 12
         Function evaluations: 2469
('time elapsed=', 0.438206246634536)
[[-0.01145566  0.36002798  0.22132933 -0.07012716  0.36645163 -0.14699988
  -0.00701157  0.07806219 -0.04311103]]
('error=', 0.0081789465690886506)


## 'CG'

In [25]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="CG",options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

         Current function value: 392.669371
         Iterations: 80
         Function evaluations: 1937
         Gradient evaluations: 175
('time elapsed=', 0.11200509155355576)
[-0.0115803   0.36259689  0.21770504 -0.0696578   0.36784854 -0.14696708
 -0.00694023  0.07745365 -0.0421006 ]
('error=', 3.7400650385306241e-07)


C:\Users\Tono\Anaconda2\lib\site-packages\ipykernel_launcher.py:6: OptimizeWarning: Unknown solver options: xtol
  


## 'BFGS'

In [26]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="BFGS",options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 12
         Function evaluations: 220
         Gradient evaluations: 20
('time elapsed=', 0.02212107768059468)
[-0.0115803   0.36259688  0.21770505 -0.06965779  0.36784855 -0.14696708
 -0.00694022  0.07745366 -0.04210063]
('error=', 3.2554707942736406e-07)


C:\Users\Tono\Anaconda2\lib\site-packages\ipykernel_launcher.py:6: OptimizeWarning: Unknown solver options: xtol
  


## 'Newton-CG'

In [27]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="Newton-CG",
               jac=ridge_reg_der,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

         Current function value: 392.669371
         Iterations: 12
         Function evaluations: 13
         Gradient evaluations: 489
         Hessian evaluations: 0
('time elapsed=', 0.031903987395935474)
[-0.0115803   0.36259694  0.21770513 -0.06965771  0.36784859 -0.14696702
 -0.00694018  0.07745367 -0.04210075]
('error=', 7.6273540328160334e-11)


## 'L-BFGS-B'

In [55]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="L-BFGS-B",        
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

('time elapsed=', 0.013529278605801665)
[-0.01157845  0.36259321  0.21770404 -0.06966662  0.36784052 -0.14696505
 -0.00694579  0.07745245 -0.04208973]
('error=', 3.0359742900718419e-05)


C:\Users\Tono\Anaconda2\lib\site-packages\ipykernel_launcher.py:8: OptimizeWarning: Unknown solver options: xtol
  


In [31]:
print res.nit
print res.fun

18
[[ 392.66937123]]


## 'TNC'

In [56]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="TNC",             
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

('time elapsed=', 0.05220225098173614)
[-0.00810599  0.34864842  0.1689608  -0.09943443  0.34379947 -0.17151509
 -0.02774523  0.06516421  0.03868742]
('error=', 0.18368018953459769)


In [33]:
print res.nit
print res.fun

11
[[ 392.66937111]]


## 'COBYLA'

In [42]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="COBYLA",
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

('time elapsed=', 0.019974708158201793)
[-0.0115149   0.36134047  0.21438375 -0.07226335  0.36583802 -0.14855338
 -0.00895643  0.07664893 -0.03621637]
('error=', 0.013686881417741169)


C:\Users\Tono\Anaconda2\lib\site-packages\ipykernel_launcher.py:7: OptimizeWarning: Unknown solver options: xtol
  import sys


In [45]:
print res.keys()
print res.fun

['status', 'nfev', 'maxcv', 'success', 'fun', 'x', 'message']
392.674195446


## 'SLSQP' 

In [46]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="SLSQP",
               jac=ridge_reg_der,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.    (Exit mode 0)
            Current function value: 392.669371103
            Iterations: 12
            Function evaluations: 45
            Gradient evaluations: 12
('time elapsed=', 0.05966298877513054)
[-0.0115809   0.36259591  0.2177053  -0.0696582   0.36784794 -0.14696801
 -0.00694075  0.07745273 -0.0420991 ]
('error=', 4.4887893733382266e-06)


C:\Users\Tono\Anaconda2\lib\site-packages\ipykernel_launcher.py:8: OptimizeWarning: Unknown solver options: xtol
  


## 'dogleg'

In [47]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="dogleg",
               jac=ridge_reg_der,hess=ridge_reg_hess,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

C:\Users\Tono\Anaconda2\lib\site-packages\scipy\optimize\_minimize.py:461: OptimizeWarning: Unknown solver options: xtol
  callback=callback, **options)


Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 1
         Function evaluations: 2
         Gradient evaluations: 2
         Hessian evaluations: 1
('time elapsed=', 0.27855396402810584)
[-0.0115803   0.36259694  0.21770513 -0.06965771  0.36784859 -0.14696702
 -0.00694018  0.07745367 -0.04210075]
('error=', 7.2355540350356572e-15)


## 'trust-ncg'

In [48]:
#We get the origin as the seed point.
beta_ls0 = np.zeros(b)
time_start = time.clock() #we obtain the time

#We use 
res = minimize(ridge_reg,beta_ls0,args=(X,Y,rho),method="trust-ncg",
               jac=ridge_reg_der,hess=ridge_reg_hess,
               options={"disp":True,"xtol":1e-10}) 
time_elapsed = (time.clock() - time_start)
print('time elapsed=',time_elapsed)
print (res.x) 
print('error=',np.linalg.norm(beta_ls_exact.T-res.x,ord=2)/np.linalg.norm(beta_ls_exact.T,ord=2))

Optimization terminated successfully.
         Current function value: 392.669371
         Iterations: 12
         Function evaluations: 13
         Gradient evaluations: 13
         Hessian evaluations: 12
('time elapsed=', 0.006294120970210315)
[-0.0115803   0.36259694  0.21770512 -0.06965772  0.3678486  -0.14696703
 -0.00694017  0.07745368 -0.04210075]
('error=', 1.6986602378050681e-08)


C:\Users\Tono\Anaconda2\lib\site-packages\scipy\optimize\_minimize.py:464: OptimizeWarning: Unknown solver options: xtol
  callback=callback, **options)


Now let's compare the results:

In [59]:
from tabulate import tabulate
print tabulate([['Nelder-Mead',1236, 406, 0.15 ], 
                ['Powell', 12, 392.6, 0.43],
                ['CG',80, 392.6, 0.11,175],
                ['BFGS', 12, 392.6, 0.02,20],
                ['Newton-CG', 12, 392.6, 0.03,489],
                ['L-BFGS-B', 18, 392.6, 0.03,489],
                ['TNC', 11, 392.6, 0.005],
                ['COBYLA', 'not available', 392.6, 0.01],
                ['SLSQP', 12,392.6, 0.05,12],
                ['dogleg', 1, 392.6, 0.27,2,1],
                ['trust-ncg', 12, 392.6, 0.006,13,12],
                ], 
               headers=['Method','Iterations', 'Function Value', 'Time', 'Gradient Evaluations','Hessian Evaluations'])

Method       Iterations       Function Value    Time    Gradient Evaluations    Hessian Evaluations
-----------  -------------  ----------------  ------  ----------------------  ---------------------
Nelder-Mead  1236                      406     0.15
Powell       12                        392.6   0.43
CG           80                        392.6   0.11                      175
BFGS         12                        392.6   0.02                       20
Newton-CG    12                        392.6   0.03                      489
L-BFGS-B     18                        392.6   0.03                      489
TNC          11                        392.6   0.005
COBYLA       not available             392.6   0.01
SLSQP        12                        392.6   0.05                       12
dogleg       1                         392.6   0.27                        2                      1
trust-ncg    12                        392.6   0.006                      13                     12


From the table, we coul say that the best ones are TNC and trust-ncg because they compute quicklier, obtaining both the solution.

**c)** Estimate the value of the regression coefficients by implementing the:

i. Gradient method.

ii. Newton method.

Consider a line search technique to improve the algorithm convergence, e.x., Armijo rule.
Compare the performance of these algorithms (number of iterations and total computational
time).

## GRADIENT METHOD (with Wolfe Conditions) 


 $\rightarrow$ From an initial iterate $x_0$

$\rightarrow$ Compute search (descent) directions $p_k=-\nabla f(x_k)$

$\rightarrow$ Far from the solution, compute a steplength $\alpha_k>0$:

&nbsp;&nbsp;&nbsp;&nbsp; $\rightarrow$ For the computation of $\alpha_k$ use Wolfe conditions, let it be $ 0< c_1 < c_2 <1$ :
  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Armijo rule : 
$f(x_k+\alpha_k p_k)\leq f(x_k)+c_1 \alpha_k\nabla f(x_k)' p_k$

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Curvature rule : 
$\nabla f(x_k+\alpha_k p_k)'p_k \geq c_2 \nabla f(x_k)' p_k$


$\rightarrow$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [89]:
(a,b)=X.shape
beta_lsg=np.zeros(b) #seed for beta
n_iter=5000 #maximum number of iterations
OF_iter=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
i=0;
tol=10000;
epsilon=1e-3;

time_start = time.clock()
while (i <= n_iter-2) and (tol>epsilon):
    i=i+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    ddirect=-grad
    
    # Here we use the Wolfe Conditions
    #######################################
    c1=0.01
    c2=0.8
    alphaMax = 1
    alphaMin = 0.00001
    alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin
    advance = False
    while advance == False:
        
        # Armijo rule
        if ( ridge_reg(beta_lsg+alpha*ddirect,X,Y,rho)<
           ridge_reg(beta_lsg,X,Y,rho)+alpha*c1*np.dot(ddirect,grad) ): 1
        else: 
            alphaMax=alpha
            alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin    
            continue   
        # Curvature rule
        if (np.dot(ddirect,ridge_reg_der(beta_lsg+alpha*ddirect,X,Y,rho)) >
           c2*np.dot(ddirect,grad) ): advance = True
        else:
            alphaMin=alpha
            alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin
    
    
    #######################################
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter[i]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.absolute((OF_iter[i]-OF_iter[i-1])/OF_iter[i]) #tolerance
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[i]=tol
    alpha_iter[i]=alpha

time_elapsed = (time.clock() - time_start) 
print('time elapsed=',time_elapsed)
print('iterations',i)
print(OF_iter[i])
print("COMPUTED BETAS:",beta_lsg)
print("EXACT BETAS",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('error=',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

table =np.zeros(shape=(2,3))
table[0,:]=[i,time_elapsed,np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2)]

('time elapsed=', 0.11545516426485847)
('iterations', 105)
392.669371101
('COMPUTED BETAS:', array([-0.01158029,  0.36259659,  0.21770432, -0.06965834,  0.36784813,
       -0.14696747, -0.00694066,  0.07745347, -0.04209932]))
('EXACT BETAS', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 0.00047821697022872446)
('error=', 3.3483089140720788e-06)


## NEWTON METHOD (with Wolfe Conditions) 


 $\rightarrow$ From an initial iterate $x_0$

$\rightarrow$ Compute search (descent) directions $p_k=-(\nabla^2 f(x_k))^{-1} \nabla f(x_k)$, whenever $\nabla^2 f(x_k)$

$\rightarrow$ Far from the solution, compute a steplength $\alpha_k>0$:

&nbsp;&nbsp;&nbsp;&nbsp; $\rightarrow$ For the computation of $\alpha_k$ use Wolfe conditions, let it be $ 0< c_1 < c_2 <1$:
  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Armijo rule : 
$f(x_k+\alpha_k p_k)\leq f(x_k)+c_1 \alpha_k\nabla f(x_k)' p_k$

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Curvature rule : 
$\nabla f(x_k+\alpha_k p_k)'p_k \geq c_2 \nabla f(x_k)' p_k$




$\rightarrow$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [80]:
(a,b)=X.shape
beta_lsn=np.zeros(b) #seed for beta
n_iter=5000 #maximum number of iterations
OF_iter=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
i=0;
tol=10000;
epsilon=1e-3;

time_start = time.clock()
while (i <= n_iter-2) and (tol>epsilon):
    i=i+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    hess= ridge_reg_hess(beta_lsn,X,Y,rho)
    ddirect= -np.dot(inv(hess),grad)
    
    # Here we use the Wolfe Conditions
    #######################################
    c1=0.01
    c2=0.8
    alphaMax = 1
    alphaMin = 0.00001
    alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin
    advance = False
    while advance == False:
        
        # Armijo rule
        if ( ridge_reg(beta_lsg+alpha*ddirect,X,Y,rho)<
           ridge_reg(beta_lsg,X,Y,rho)+alpha*c1*np.dot(ddirect,grad) ): 1
        else: 
            alphaMax=alpha
            alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin    
            continue   
        # Curvature rule
        if (np.dot(ddirect,ridge_reg_der(beta_lsg+alpha*ddirect,X,Y,rho)) >
           c2*np.dot(ddirect,grad) ): advance = True
        else:
            alphaMin=alpha
            alpha = np.random.rand(1)[0]*(alphaMax-alphaMin) + alphaMin
    
    
    #######################################
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter[i]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.absolute((OF_iter[i]-OF_iter[i-1])/OF_iter[i]) #tolerance
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[i]=tol
    alpha_iter[i]=alpha

time_elapsed = (time.clock() - time_start) 
print('time elapsed=',time_elapsed)
print('iterations',i)
print(OF_iter[i])
print("COMPUTED BETAS:",beta_lsg)
print("EXACT BETAS",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('error=',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

table =np.zeros(shape=(2,3))
table[0,:]=[i,time_elapsed,np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2)]

('time elapsed=', 0.0018121474322470021)
('iterations', 2)
392.6693711
('COMPUTED BETAS:', array([-0.0115803 ,  0.36259688,  0.21770499, -0.06965781,  0.36784851,
       -0.14696709, -0.00694026,  0.07745364, -0.04210053]))
('EXACT BETAS', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 0.00051853845989666765)
('error=', 5.2917180142352911e-07)


**Comparison**: Gradient Method needed 0.11 seconds and 105 iterations, while Newton Method needed 0.001 seconds and 2 iterations. We would say that Newton Method is better in that case.

**d)** Estimate the value of the regression coefficients by implementing the:

i. Coordinate gradient method.

ii. Stochastic gradient method.

## Coordinate gradient method

$\rightarrow$ From an initial iterate $x_0$

$\rightarrow$ Compute search (descent) directions. Find $i_k \in {1,...,p}$ such that $\nabla_{i_k} f(x_k)$ is maximum. $p_k=- \nabla_{i_k} f(x_k) e_{i_k}$

$\rightarrow$ Far from the solution, compute a steplength $\alpha_k>0$:

$\rightarrow$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.

In [105]:
(a,b)=X.shape
beta_lsg=np.zeros(b) #initial value for beta

alpha=0.0001
n_iter=5000 #maximum number of iterations
OF_iter=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
i=0;
tol=10000;
epsilon=1e-3;

time_start = time.clock()
while (i <= n_iter-2) and (tol>epsilon):
    i=i+1
    grad= ridge_reg_der(beta_lsg,X,Y,rho)
    ddirect=-np.eye(1,len(grad),np.argmax(abs(grad)))*grad
    
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter[i]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[i]=tol
    alpha_iter[i]=alpha
    

time_elapsed = (time.clock() - time_start) 
print('time elapsed=',time_elapsed)
print('iterations',i)
print(OF_iter[i])
print("COMPUTED BETAS:",beta_lsg)
print("EXACT BETAS",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('error=',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

('time elapsed=', 0.2660486850181769)
('iterations', 2050)
392.669371103
('COMPUTED BETAS:', array([[-0.0115802 ,  0.36259574,  0.2177029 , -0.06965961,  0.36784715,
        -0.14696837, -0.00694162,  0.07745311, -0.04209661]]))
('EXACT BETAS', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 0.00099418722578349292)
('error=', 9.7917516091174429e-06)


The time needed was 0.26 seconds and the number of iterations was 2050.

## Stochastic gradient method

$\rightarrow$ From an initial iterate $x_0$

$\rightarrow$ Compute search (descent) directions. Choose randomly $j_k \in {1,...,n}$. $p_k=- \nabla f_{j_k}(x_k) $

$\rightarrow$ Far from the solution, compute a steplength $\alpha_k>0$:

$\rightarrow$ Movement:
$$x_{k+1} = x_k + \alpha_k\ p_k$$
Until convergence to a local solution.



In [146]:
import random

In [149]:
#Random Gradient:

def ridge_reg_der2(beta_ls,X,Y,rho):
    j=int(random.uniform(1,len(Y))) #we choose one individual randomly
    beta_ls=np.matrix(beta_ls)
    pp=-2*(Y[j]-np.dot(X[j,:],beta_ls.T)).T*X[j,:]+2*rho*beta_ls
    aa=np.squeeze(np.asarray(pp))
    return aa

In [152]:
(a,b)=X.shape
beta_lsg=np.zeros(b) #initial value for beta

alpha=0.0001
n_iter=5000 #maximim number of iterations
OF_iter=np.zeros(n_iter)
tol_iter=np.zeros(n_iter)
alpha_iter=np.zeros(n_iter)
i=0;
tol=10000;
epsilon=1e-3;

time_start = time.clock()
while (i <= n_iter-2) and (tol>epsilon):
    i=i+1
    grad= ridge_reg_der2(beta_lsg,X,Y,rho)
    ddirect=-grad
    beta_lsg= beta_lsg +alpha*ddirect
    OF_iter[i]=ridge_reg(beta_lsg, X, Y,rho)
    tol=np.linalg.norm(grad,ord=2)
    tol_iter[i]=tol
    alpha_iter[i]=alpha
    

time_elapsed = (time.clock() - time_start) 
print('time elapsed=',time_elapsed)
print('iterations',i)
print(OF_iter[i])
print("COMPUTED BETAS:",beta_lsg)
print("EXACT BETAS",np.transpose(beta_ls_exact))
print('Tolerance=',tol)
print('error=',np.linalg.norm(np.transpose(beta_ls_exact)-beta_lsg,ord=2)/np.linalg.norm(beta_lsg,ord=2))

('time elapsed=', 0.5177432028667681)
('iterations', 4999)
537.462295573
('COMPUTED BETAS:', array([-0.01792542,  0.1786421 ,  0.15682673, -0.09972942,  0.14898236,
       -0.1007175 , -0.02768812,  0.09917452,  0.10715754]))
('EXACT BETAS', array([[-0.0115803 ,  0.36259694,  0.21770513, -0.06965771,  0.36784859,
        -0.14696702, -0.00694018,  0.07745367, -0.04210075]]))
('Tolerance=', 0.51220819901803927)
('error=', 0.96006154468567928)


We can observe that choosing an individual randomly and calculating the gradient for the function defined for this individual gives us 'COMPUTED BETAS' quite far from 'EXACT BETAS'. Also we have exceeded the maximum number of iterations.